### PR TITLE
docs: add DeepSeek-R1 training dynamics and GRPO example

### PR DESCRIPTION
**Pull request by JenWei0312**
Adds a detailed entry for the DeepSeek-R1 paper (arXiv:2501.12948) to the TRL paper index.

Part of https://github.com/huggingface/trl/issues/4407.

# What does this PR do?

Fixes # 4407

This PR adds the DeepSeek-R1 paper (2501.12948) to the paper index as requested in the tracking issue.

Key additions:
- Breakdown of the four-stage pipeline: Cold Start, Reasoning-RL, Rejection Sampling/SFT, and Full-Scenario RL.
- Explanation of the role of rule-based vs. neural reward models in the final alignment stage.
- Documentation of the distillation strategy used to empower smaller dense models with R1's reasoning patterns.
- An updated GRPOTrainer example configuration including vLLM integration and multi-stage reward logic.

## Before submitting
- [x ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ x] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by JenWei0312** [comment]
Hi @sergiopaniego, as discussed in #4407, this PR is ready for your review! I've included the DeepSeek-R1 multi-stage pipeline details and a GRPOTrainer example. Looking forward to your feedback. 🔥

File: docs/source/paper_index.md [hunk 1]
```diff
@@ -34,6 +34,50 @@ trainer = GRPOTrainer(
...
+
+**📜 Paper**: https://huggingface.co/papers/2501.12948
+
+DeepSeek-R1 achieves reasoning performance comparable to OpenAI-o1 through a multi-stage pipeline that transitions from pure reinforcement learning (RL) to a refined, human-aligned model. Unlike its predecessor, DeepSeek-R1-Zero, which used pure RL on a base model, R1 follows a structured four-stage evolution:
+1. Cold Start: The base model is fine-tuned on a small set of high-quality, long Chain-of-Thought (CoT) data to provide a stable starting point.
+2. Reasoning-Oriented RL: Large-scale RL is applied to enhance performance in math, coding, and logic, using rule-based rewards and a language consistency reward to reduce language mixing.
+3. Rejection Sampling & SFT: The RL checkpoint generates 600k reasoning samples via rejection sampling, which are combined with 200k non-reasoning (general) samples to create a new dataset for a second round of Supervised Fine-Tuning.
+4. RL for all Scenarios: A final RL stage aligns the model with human preferences (helpfulness and harmlessness) across all domains while maintaining reasoning strength.
+
+Distillation: Empowering Small Models
```
Comment (sergiopaniego) [inline docs/source/paper_index.md:47]:
```suggestion
Distillation: Empowering Small Models

```

File: docs/source/paper_index.md [hunk 2]
```diff
@@ -34,6 +34,50 @@ trainer = GRPOTrainer(
...
+    dataset_name="open-r1/Mixture-of-Thoughts",
+    learning_rate=4.0e-5,
+    max_prompt_length=4096,
+    max_completion_length=32768, # Support for long Chain-of-Thought
+    num_generations=16,          # Sample 16 outputs per prompt for group relative advantage
+    beta=0.001,                  # KL coefficient
+    use_vllm=True,               # Use vLLM backend for accelerated rollout generation
+    vllm_device="cuda:1",        # Optional: colocate vLLM on a specific device
+    report_to="wandb",
+)
```
Comment (sergiopaniego) [inline docs/source/paper_index.md:65]:
```suggestion
training_args = GRPOConfig(
    learning_rate=4.0e-5,
    max_prompt_length=4096,
    max_completion_length=32768, # Support for long Chain-of-Thought
    num_generations=16,          # Sample 16 outputs per prompt for group relative advantage
    beta=0.001,                  # KL coefficient
    use_vllm=True,               # Use vLLM backend for accelerated rollout generation
)
```
